src/agent/coverage/src/allowlist.rs

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use anyhow::Result;
+use regex::{Regex, RegexSet};
+use std::path::Path;
+
+#[derive(Clone, Debug, Default)]
+pub struct TargetAllowList {
+    pub functions: AllowList,
+    pub modules: AllowList,
+    pub source_files: AllowList,
+}
+
+impl TargetAllowList {
+    pub fn new(modules: AllowList, source_files: AllowList) -> Self {
+        // Allow all.
+        let functions = AllowList::default();
+
+        Self {
+            functions,
+            modules,
+            source_files,
+        }
+    }
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct AllowList {
+    allow: Option<RegexSet>,
+    deny: Option<RegexSet>,
+}
+
+impl AllowList {
+    pub fn new(allow: impl Into<Option<RegexSet>>, deny: impl Into<Option<RegexSet>>) -> Self {
```

Comment:
Is there a difference between the impl trait argument and defining a generic T in the function?

I'm asking because the types for allow and deny are the same so they could be factored out together:

```rust
pub fn new<T: Into<Option<RegexSet>>>(allow: T, deny: T) -> Self {
```

Reply:
There is a difference when there are multiple `impl Trait` arguments. Each `impl Trait` introduces a fresh type variable. So your example is not equivalent, because it requires that the arguments have the same concrete type at each call site.

This would prevent what we're trying to achieve, which is simulating optional arguments. As written, we could call `AllowList::new(regex_set, None)`, where one argument has the non-optional type `RegexSet`, but where the second argument has type `Option<RegexSet>`.

Reply:
I ended up changing the signature anyway 😄.